drivers: flash: flash page layout returns error if pages_size null

FRASTM opened the pull request:
Return the error code if the layout->pages_size is null instead of going to a "divide by zero" fault
This can occur when CONFIG_FLASH_PAGE_LAYOUT is set and read out of range.

Run samples/drivers/flash_shell on a stm32u575 nucleo board  and read flash out of range :
```
flash read 0x2a0000 0x100
Read ERROR!
[00:00:13.919,000] <err> flash_stm32: Read range invalid. Offset: 2752512, len: 16

```

With the page layout **CONFIG_FLASH_PAGE_LAYOUT=y**
```
flash read 0x2a0000 0x100
[00:00:20.887,000] <err> os: ***** USAGE FAULT *****
[00:00:20.887,000] <err> os:   Division by zero
[00:00:20.887,000] <err> os: r0/a1:  0x200010f4  r1/a2:  0x00200000  r2/a3:  0x00000100
[00:00:20.887,000] <err> os: r3/a4:  0x000a0000 r12/ip:  0x00000100 r14/lr:  0x0800c021
[00:00:20.887,000] <err> os:  xpsr:  0x21000000
[00:00:20.887,000] <err> os: Faulting instruction address (r15/pc): 0x0800c060
[00:00:20.887,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:20.887,000] <err> os: Current thread: 0x20000700 (shell_uart)
[00:00:20.950,000] <err> os: Halting system

```

This change avoid the Usage Fault 

Signed-off-by: Francois Ramu <francois.ramu@st.com>

de-nordic commented on drivers/flash/flash_page_layout.c:
```diff
@@ -23,7 +23,10 @@ static int flash_get_page_info(const struct device *dev, off_t offs,
-		if (offs == 0) {
+		if (info->size == 0) {
+			/* Avoid dividing by 0 in the next operation */
+			return -EINVAL;
```
You can just break out here, the default return does the same thing.

Laczen commented:
I don't think this should be checked, when `CONFIG_FLASH_PAGE_LAYOUT` is selected that information should not be zero. If page layout is not supported a compile error should be generated when `CONFIG_FLASH_PAGE_LAYOUT` is selected.

de-nordic reviewed (commented):
OK, so I do not understand something here.
The `flash_stm32_read` depends on `flash_stm32_valid_range` to validate range.
The `flash_stm32_valid_range` for stm32u575 , which, I guess, is defined here drivers/flash/flash_stm32l5_u5.c requires `flash_stm32_range_exists`, which is only provided with drivers/flash/flash_stm32.h when `CONFIG_FLASH_PAGE_LAYOUT` is y.
How does this work correctly when `CONFIG_FLASH_PAGE_LAYOUT` is n?
How does it compile when `CONFIG_FLASH_PAGE_LAYOUT` is n?

carlescufi commented:
@FRASTM @erwango can you reply to @de-nordic's review comments?

Laczen commented:
@FRASTM, although I think it is good to check against null pointers it seems like this is hiding a configuration error.

FRASTM commented:
> @FRASTM @erwango can you reply to @de-nordic's review comments?

The stm32u5 defines the CONFIG_FLASH_PAGE_LAYOUT
Then flash_stm32_page_layout() in the [flash_stm32l5_u5.c ](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_stm32l5_u5.c)   for that target gives a layout table of one element but defines a table of 3. It this case the array size of the stm32_flash_layout[] could be reduced o 1.
-->Instead of checking the "div by zero",  I can fix in this way : 
```
diff --git a/drivers/flash/flash_stm32l5_u5.c b/drivers/flash/flash_stm32l5_u5.c
index ef72ad662a..05714175be 100644
--- a/drivers/flash/flash_stm32l5_u5.c
+++ b/drivers/flash/flash_stm32l5_u5.c
@@ -404,6 +404,11 @@ void flash_stm32_page_layout(const struct device *dev,
 			stm32_flash_layout[0].pages_size = FLASH_PAGE_SIZE_128_BITS;
 #endif /* CONFIG_SOC_SERIES_STM32L5X */
 		}
+		/*
+		 * In this case the stm32_flash_layout table has one single element
+		 * when read by the flash_get_page_info()
+		 */
+		*layout_size = 1;
 	}
 
 }
```

erwango commented:
> changing at runtime from dual bank to single bank 

@Laczen Do you have a use case in mind for this?

FRASTM committed:
drivers: flash: stm32 page layout size differs in stm32u5 or stm32l5

Adjust the size of the stm32_flash_layout[] table depending on the
bank configuration of the stm32u5 or stm32l5 devices.
That will avoid div by zero error in flash_get_page_info()
if the layout_size is not correct.
Assign the *layout_size only once with the correct value : 3 or 1.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

Laczen commented:
> > changing at runtime from dual bank to single bank
> 
> @Laczen Do you have a use case in mind for this?

No, it might never be allowed, but to fix the problem the `stm_flash_page_layout` needs to be either specific for each use (dual vs single bank) or made the same for both cases. The first seems like a obvious solution, the second just jumped into my head so I proposed this alternative approach. 😀

FRASTM commented:
With this new commit, the array size is reflecting the layout for stm32l5 or stm32u5 in case of one bank config
That will avoid the div by zero error in the flash_get_page_info() as the layout _size is 3 or 1.

Note that this is like the stm32h7 does in the flash-stm32h7x.c

FRASTM commented:
> > > changing at runtime from dual bank to single bank
> > 
> > 
> > @Laczen Do you have a use case in mind for this?
> 
> No, it might never be allowed, but to fix the problem the `stm_flash_page_layout` needs to be either specific for each use (dual vs single bank) or made the same for both cases. The first seems like a obvious solution, the second just jumped into my head so I proposed this alternative approach. grinning

If the layout is for one flash area, why giving three ? 
The structure just informed about nb of pages of the same size. Modifying the layout at runtime is another story

FRASTM commented:
rebase on fae717305bc91e9e5d62dbcda18a77312e18a624
